### PR TITLE
FINERACT-2096: Fix duplicate key value exception

### PR DESCRIPTION
**Pull request by alberto-art3ch**
## Description

Fix duplicate key value exception to be managed as `IdempotentCommandProcessUnderProcessingException` with HTTP code 425

[FINERACT-2096](https://issues.apache.org/jira/browse/FINERACT-2096)

Screenshot 2024-06-13 at 7 13 51 a m

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-core/src/main/java/org/apache/fineract/commands/service/CommandSourceService.java [hunk 1]
```diff
@@ -64,8 +66,12 @@ public CommandSource saveInitialSameTransaction(CommandWrapper wrapper, JsonComm
 
     @NotNull
     private CommandSource saveInitial(CommandWrapper wrapper, JsonCommand jsonCommand, AppUser maker, String idempotencyKey) {
-        CommandSource initialCommandSource = getInitialCommandSource(wrapper, jsonCommand, maker, idempotencyKey);
-        return commandSourceRepository.saveAndFlush(initialCommandSource);
+        try {
+            CommandSource initialCommandSource = getInitialCommandSource(wrapper, jsonCommand, maker, idempotencyKey);
+            return commandSourceRepository.saveAndFlush(initialCommandSource);
+        } catch (JpaSystemException jse) {
```
Comment (adamsaghy) [inline fineract-core/src/main/java/org/apache/fineract/commands/service/CommandSourceService.java:72]: This is not checking whether the issue was with the duplicate idempotency key.. Please de the proper checks! We want to throw IdempotentCommandProcessUnderProcessingException **only** if it failed with idempotency key duplication error!

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please kindly check my review!

File: fineract-core/src/main/java/org/apache/fineract/commands/service/CommandSourceService.java (hunk 1 above)
Comment (alberto-art3ch) [reply fineract-core/src/main/java/org/apache/fineract/commands/service/CommandSourceService.java:72]: Done! Updated to throw the exception only when we have a SQLIntegrityConstraintViolationException

**Comment by alberto-art3ch** [comment]
> Please kindly check my review!

Changes applied please review

File: fineract-core/src/main/java/org/apache/fineract/commands/service/CommandSourceService.java [hunk 2]
```diff
@@ -64,8 +66,15 @@ public CommandSource saveInitialSameTransaction(CommandWrapper wrapper, JsonComm
...
+            if (jse.getRootCause().getMessage().contains("UNIQUE_PORTFOLIO_COMMAND_SOURCE")) {
```
Comment (adamsaghy) [inline fineract-core/src/main/java/org/apache/fineract/commands/service/CommandSourceService.java]: Please use null checks! Root cause and message both can be null!

Comment (alberto-art3ch) [reply fineract-core/src/main/java/org/apache/fineract/commands/service/CommandSourceService.java]: Done! null checks added

File: fineract-core/src/main/java/org/apache/fineract/commands/service/CommandSourceService.java [hunk 3]
```diff
@@ -64,8 +66,16 @@ public CommandSource saveInitialSameTransaction(CommandWrapper wrapper, JsonComm
...
     @NotNull
     private CommandSource saveInitial(CommandWrapper wrapper, JsonCommand jsonCommand, AppUser maker, String idempotencyKey) {
-        CommandSource initialCommandSource = getInitialCommandSource(wrapper, jsonCommand, maker, idempotencyKey);
-        return commandSourceRepository.saveAndFlush(initialCommandSource);
+        try {
+            CommandSource initialCommandSource = getInitialCommandSource(wrapper, jsonCommand, maker, idempotencyKey);
+            return commandSourceRepository.saveAndFlush(initialCommandSource);
+        } catch (JpaSystemException jse) {
+            final String message = (jse.getRootCause() != null) ? jse.getRootCause().getMessage() : null;
+            if (message != null && message.contains("UNIQUE_PORTFOLIO_COMMAND_SOURCE")) {
```
Comment (adamsaghy) [inline fineract-core/src/main/java/org/apache/fineract/commands/service/CommandSourceService.java]: I reckon we need to slightly modify to work with MariaDB and Postgres as well:
`message.toLowerCase().contains("unique_portfolio_command_source")`

**Review by adamsaghy — CHANGES_REQUESTED** [review]
Please kindly check my review!

File: fineract-core/src/main/java/org/apache/fineract/commands/service/CommandSourceService.java (hunk 3 above)
Comment (alberto-art3ch) [reply fineract-core/src/main/java/org/apache/fineract/commands/service/CommandSourceService.java]: Done!

**Review by adamsaghy — APPROVED** [review]
LGTM

**Pull request merged**